# **Title:**
#        **ITSM Improvement through Machine Learning: Enhancing Incident Management at ABC Tech.**

# **PROBLEM** **OF** **STATEMENT:**

*   Predicting High Priority Tickets: To predict priority 1 & 2 tickets, so that they can take preventive measures or fix the problem before it surfaces.

*   Forecast the incident volume in different fields , quarterly and annual. So that they can be better prepared with resources and technology planning.

*    Auto tag the tickets with right priorities and right departments so that reassigning and related delay can be reduced.

*    Predict RFC (Request for change) and possible failure / misconfiguration of ITSM assets.









# **Description:**
## **Business Case Description:**

ABC Tech is an established mid-sized organization operating in the IT-enabled business sector for over a decade. They manage a significant volume of IT incidents and tickets, averaging between 22,000 to 25,000 per year. ABC Tech follows best practices in IT Service Management (ITSM), including incident management, problem management, change management, and configuration management processes. These ITIL practices have matured over time, reaching a high level of process maturity.

Recently, ABC Tech conducted an audit that indicated that further improvement initiatives in their ITSM processes may not provide a sufficient return on investment (ROI). Despite their mature processes, customer feedback from recent surveys has revealed that incident management, in particular, is rated poorly, suggesting there is room for enhancement.

In response to these challenges, ABC Tech's management has decided to explore the potential of machine learning (ML) to enhance their ITSM processes. After attending a Machine Learning conference focused on IT Service Management (ITSM), they identified four key areas where ML can contribute to improving ITSM processes within the organization:

1. **Predicting High Priority Tickets:** ABC Tech aims to develop an ML model that can predict high-priority tickets, specifically those categorized as priority 1 and 2. This prediction will allow them to take proactive measures to address issues or incidents before they escalate.

2. **Forecasting Incident Volume:** The organization plans to use ML to forecast the incident volume in different fields on a quarterly and annual basis. This predictive capability will help them better allocate resources and plan for the required technology upgrades.

3. **Auto-Tagging Tickets:** ABC Tech intends to implement a text classification ML model to automatically assign correct priorities and departments to incoming tickets. This automation will reduce reassignment and related delays in ticket handling.

4. **Predicting RFC and ITSM Asset Misconfigurations:** The organization aims to create predictive models for Request for Change (RFC) and detect potential failures or misconfigurations in ITSM assets. Identifying these issues in advance will help in preventing disruptions and improving overall ITSM asset management.

The dataset that ABC Tech plans to use for these ML initiatives comprises a total of approximately 46,000 records spanning the years 2012, 2013, and 2014. The data is stored in a MySQL database with read-only access, and the relevant connection details are provided.

Here's a summary of some key fields in the dataset:

- CI_Name: Configuration Item Name
- CI_Cat: Configuration Item Category
- CI_Subcat: Configuration Item Subcategory
- Incident_ID: Incident Identifier
- Status: Incident Status (e.g., Closed)
- Impact: Impact level
- Urgency: Urgency level
- Priority: Incident Priority
- Category: Incident Category
- KB_number: Knowledge Base Number
- Alert_Status: Alert Status
- No_of_Reassignments: Number of Reassignments
- Open_Time: Incident Open Time
- Reopen_Time: Incident Reopen Time
- Resolved_Time: Incident Resolved Time
- Close_Time: Incident Close Time
- Handle_Time_hrs: Time taken to handle the incident (in hours)
- Closure_Code: Code for incident closure
- No_of_Related_Interactions: Number of Related Interactions
- Related_Interaction: Identifier for Related Interaction
- No_of_Related_Incidents: Number of Related Incidents
- No_of_Related_Changes: Number of Related Changes
- Related_Change: Identifier for Related Change


## **Import Basic Libraries:**

In [22]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

## **Load Data:**

In [23]:
data=pd.read_csv("/content/ITSM_data.csv")

In [24]:
#data=pd.set_option("display.max_columns",None)
#data=pd.set_option("display.expand_frame_repr",False)

In [25]:
data

CI_Name          CI_Cat                 CI_Subcat        WBS  \
0      SUB000508  subapplication     Web Based Application  WBS000162   
1      WBA000124     application     Web Based Application  WBS000088   
2      DTA000024     application       Desktop Application  WBS000092   
3      WBA000124     application     Web Based Application  WBS000088   
4      WBA000124     application     Web Based Application  WBS000088   
...          ...             ...                       ...        ...   
46601  SBA000464     application  Server Based Application  WBS000073   
46602  SBA000461     application  Server Based Application  WBS000073   
46603  LAP000019        computer                    Laptop  WBS000091   
46604  WBA000058     application     Web Based Application  WBS000073   
46605  DCE000077        hardware       DataCenterEquipment  WBS000267   

      Incident_ID  Status Impact Urgency  Priority  number_cnt  ...  \
0       IM0000004  Closed      4       4       4.0    0.601292  ...   
1       IM0000005  Closed      3       3       3.0    0.415050  ...   
2       IM0000006  Closed     NS       3       NaN    0.517551  ...   
3       IM0000011  Closed      4       4       4.0    0.642927  ...   
4       IM0000012  Closed      4       4       4.0    0.345258  ...   
...           ...     ...    ...     ...       ...         ...  ...   
46601   IM0047053  Closed      4       4       4.0    0.231896  ...   
46602   IM0047054  Closed      4       4       4.0    0.805153  ...   
46603   IM0047055  Closed      5       5       5.0    0.917466  ...   
46604   IM0047056  Closed      4       4       4.0    0.701278  ...   
46605   IM0047057  Closed      3       3       3.0    0.902320  ...   

           Reopen_Time     Resolved_Time        Close_Time  Handle_Time_hrs  \
0                  NaN   4/11/2013 13:50   4/11/2013 13:51   3,87,16,91,111   
1      2/12/2013 12:31   2/12/2013 12:36   2/12/2013 12:36   4,35,47,86,389   
2                  NaN  13-01-2014 15:12  13-01-2014 15:13   4,84,31,19,444   
3                  NaN  14-11-2013 09:31  14-11-2013 09:31   4,32,18,33,333   
4                  NaN   8/11/2013 13:55   8/11/2013 13:55   3,38,39,03,333   
...                ...               ...               ...              ...   
46601              NaN  31-03-2014 16:29  31-03-2014 16:29            0,095   
46602              NaN  31-03-2014 15:29  31-03-2014 15:29      0,428333333   
46603              NaN  31-03-2014 15:32  31-03-2014 15:32      0,071666667   
46604              NaN  31-03-2014 15:42  31-03-2014 15:42      0,116944444   
46605              NaN  31-03-2014 22:47  31-03-2014 22:47      0,586388889   

                       Closure_Code No_of_Related_Interactions  \
0                             Other                        1.0   
1                          Software                        1.0   
2      No error - works as designed                        1.0   
3                    Operator error                        1.0   
4                             Other                        1.0   
...                             ...                        ...   
46601                         Other                        1.0   
46602                    User error                        1.0   
46603                      Hardware                        1.0   
46604                      Software                        1.0   
46605                      Hardware                        1.0   

      Related_Interaction No_of_Related_Incidents No_of_Related_Changes  \
0               SD0000007                     2.0                   NaN   
1               SD0000011                     1.0                   NaN   
2               SD0000017                     NaN                   NaN   
3               SD0000025                     NaN                   NaN   
4               SD0000029                     NaN                   NaN   
...                   ...                     ...                   ...   
46601       

In [26]:
data.drop(["CI_Name", "WBS", "Incident_ID", "KB_number", "Related_Interaction","Related_Change"], axis=1, inplace=True)

# **Domain Analysis:**
Certainly! Domain analysis is an essential step in any data science or machine learning project. It involves gaining a deep understanding of the domain-specific aspects of the problem you're trying to solve. In this case, we're analyzing the domain of IT Service Management (ITSM) within the context of ABC Tech's business case. Here's a domain analysis for ITSM:

**1. IT Service Management (ITSM):**
   - **Definition:** IT Service Management (ITSM) refers to a set of practices and processes used by organizations to design, deliver, manage, and improve IT services for their customers and end-users.
   - **Importance:** ITSM ensures that IT services are aligned with business goals, reliable, and efficiently delivered, leading to enhanced customer satisfaction and business performance.

**2. Incident Management:**
   - **Definition:** Incident Management is a core ITSM process that involves identifying, categorizing, prioritizing, and resolving incidents to restore normal service operations as quickly as possible.
   - **Challenges:** Common challenges in incident management include handling a high volume of incidents, determining incident priorities, minimizing response times, and reducing the impact on end-users.

**3. Priority in ITSM:**
   - **Definition:** Priority is a classification system used to categorize incidents based on their severity and impact on business operations. In ITIL (IT Infrastructure Library) framework, there are typically four priority levels: Priority 1 (Critical), Priority 2 (High), Priority 3 (Medium), and Priority 4 (Low).
   - **Importance:** Prioritizing incidents helps organizations allocate resources effectively and respond to critical issues promptly.

**4. ITIL Framework:**
   - **Definition:** ITIL is a widely adopted framework for ITSM that provides best practices and guidelines for managing IT services, including incident management, problem management, change management, and configuration management.
   - **Maturity Levels:** ITIL processes can mature over time, starting from ad-hoc practices and progressing to well-defined, controlled, and optimized processes. A mature ITIL framework leads to improved service quality and efficiency.

**5. Machine Learning in ITSM:**
   - **Application:** Machine learning can be applied to ITSM processes to predict incidents, automate ticket classification, forecast resource needs, and detect anomalies or misconfigurations in IT assets.
   - **Benefits:** ML can enhance incident response, reduce manual workload, improve service quality, and proactively identify issues before they impact operations.

**6. Data Considerations:**
   - **Data Sources:** Data for ITSM projects typically come from various sources, including incident tracking systems, service logs, configuration databases, and customer feedback.
   - **Data Quality:** Data quality is critical, as inaccurate or incomplete data can lead to incorrect predictions and decisions.
   - **Privacy and Compliance:** ITSM data often contains sensitive information, so privacy and compliance with data protection regulations (e.g., GDPR) are crucial considerations.

**7. Metrics and KPIs:**
   - **Key Metrics:** Key Performance Indicators (KPIs) in ITSM include Mean Time to Resolve (MTTR), Incident Response Time, Incident Closure Rate, and Customer Satisfaction.
   - **Monitoring:** Continuous monitoring of KPIs helps organizations assess the effectiveness of their ITSM processes and make improvements.

**8. Stakeholders:**
   - **Internal Stakeholders:** IT teams, IT managers, incident responders, change management teams.
   - **External Stakeholders:** End-users, customers, auditors (for compliance).

**9. Challenges:**
   - **Scalability:** Handling a large volume of incidents efficiently can be challenging.
   - **Data Integration:** Integrating data from disparate sources can be complex.
   - **Change Management:** Implementing ML solutions may require organizational and cultural changes.


## **Basic Checks:**

In [27]:
data.head()

CI_Cat              CI_Subcat  Status Impact Urgency  Priority  \
0  subapplication  Web Based Application  Closed      4       4       4.0   
1     application  Web Based Application  Closed      3       3       3.0   
2     application    Desktop Application  Closed     NS       3       NaN   
3     application  Web Based Application  Closed      4       4       4.0   
4     application  Web Based Application  Closed      4       4       4.0   

   number_cnt                 Category Alert_Status  No_of_Reassignments  \
0    0.601292                 incident       closed                 26.0   
1    0.415050                 incident       closed                 33.0   
2    0.517551  request for information       closed                  3.0   
3    0.642927                 incident       closed                 13.0   
4    0.345258                 incident       closed                  2.0   

          Open_Time      Reopen_Time     Resolved_Time        Close_Time  \
0    5/2/2012 13:32              NaN   4/11/2013 13:50   4/11/2013 13:51   
1   12/3/2012 15:44  2/12/2013 12:31   2/12/2013 12:36   2/12/2013 12:36   
2  29-03-2012 12:36              NaN  13-01-2014 15:12  13-01-2014 15:13   
3  17-07-2012 11:49              NaN  14-11-2013 09:31  14-11-2013 09:31   
4   10/8/2012 11:01              NaN   8/11/2013 13:55   8/11/2013 13:55   

  Handle_Time_hrs                  Closure_Code  No_of_Related_Interactions  \
0  3,87,16,91,111                         Other                         1.0   
1  4,35,47,86,389                      Software                         1.0   
2  4,84,31,19,444  No error - works as designed                         1.0   
3  4,32,18,33,333                Operator error                         1.0   
4  3,38,39,03,333                         Other                         1.0   

   No_of_Related_Incidents  No_of_Related_Changes  
0                      2.0                    NaN  
1                      1.0                    NaN  
2                      NaN                    NaN  
3                      NaN                    NaN  
4                      NaN                    NaN

In [28]:
data.tail()

CI_Cat                 CI_Subcat  Status Impact Urgency  Priority  \
46601  application  Server Based Application  Closed      4       4       4.0   
46602  application  Server Based Application  Closed      4       4       4.0   
46603     computer                    Laptop  Closed      5       5       5.0   
46604  application     Web Based Application  Closed      4       4       4.0   
46605     hardware       DataCenterEquipment  Closed      3       3       3.0   

       number_cnt  Category Alert_Status  No_of_Reassignments  \
46601    0.231896  incident       closed                  0.0   
46602    0.805153  incident       closed                  0.0   
46603    0.917466  incident       closed                  0.0   
46604    0.701278  incident       closed                  0.0   
46605    0.902320  incident       closed                  0.0   

              Open_Time Reopen_Time     Resolved_Time        Close_Time  \
46601  31-03-2014 16:23         NaN  31-03-2014 16:29  31-03-2014 16:29   
46602  31-03-2014 15:03         NaN  31-03-2014 15:29  31-03-2014 15:29   
46603  31-03-2014 15:28         NaN  31-03-2014 15:32  31-03-2014 15:32   
46604  31-03-2014 15:35         NaN  31-03-2014 15:42  31-03-2014 15:42   
46605  31-03-2014 17:24         NaN  31-03-2014 22:47  31-03-2014 22:47   

      Handle_Time_hrs Closure_Code  No_of_Related_Interactions  \
46601           0,095        Other                         1.0   
46602     0,428333333   User error                         1.0   
46603     0,071666667     Hardware                         1.0   
46604     0,116944444     Software                         1.0   
46605     0,586388889     Hardware                         1.0   

       No_of_Related_Incidents  No_of_Related_Changes  
46601                      NaN                    NaN  
46602                      NaN                    NaN  
46603                      NaN                    NaN  
46604                      NaN                    NaN  
46605                      NaN                    NaN

In [29]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46606 entries, 0 to 46605
Data columns (total 19 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   CI_Cat                      46495 non-null  object 
 1   CI_Subcat                   46495 non-null  object 
 2   Status                      46606 non-null  object 
 3   Impact                      46606 non-null  object 
 4   Urgency                     46606 non-null  object 
 5   Priority                    45226 non-null  float64
 6   number_cnt                  46606 non-null  float64
 7   Category                    46606 non-null  object 
 8   Alert_Status                46606 non-null  object 
 9   No_of_Reassignments         46605 non-null  float64
 10  Open_Time                   46606 non-null  object 
 11  Reopen_Time                 2284 non-null   object 
 12  Resolved_Time               44826 non-null  object 
 13  Close_Time                  466

In [30]:
data.dtypes

CI_Cat                         object
CI_Subcat                      object
Status                         object
Impact                         object
Urgency                        object
Priority                      float64
number_cnt                    float64
Category                       object
Alert_Status                   object
No_of_Reassignments           float64
Open_Time                      object
Reopen_Time                    object
Resolved_Time                  object
Close_Time                     object
Handle_Time_hrs                object
Closure_Code                   object
No_of_Related_Interactions    float64
No_of_Related_Incidents       float64
No_of_Related_Changes         float64
dtype: object

In [31]:
data.columns

Index(['CI_Cat', 'CI_Subcat', 'Status', 'Impact', 'Urgency', 'Priority',
       'number_cnt', 'Category', 'Alert_Status', 'No_of_Reassignments',
       'Open_Time', 'Reopen_Time', 'Resolved_Time', 'Close_Time',
       'Handle_Time_hrs', 'Closure_Code', 'No_of_Related_Interactions',
       'No_of_Related_Incidents', 'No_of_Related_Changes'],
      dtype='object')

In [32]:
cat_data=data.select_dtypes(include=["O"]).columns
cat_data

Index(['CI_Cat', 'CI_Subcat', 'Status', 'Impact', 'Urgency', 'Category',
       'Alert_Status', 'Open_Time', 'Reopen_Time', 'Resolved_Time',
       'Close_Time', 'Handle_Time_hrs', 'Closure_Code'],
      dtype='object')

In [33]:
for i in cat_data:
  print(i,data[i].unique())
  print(data[i].value_counts())
  print("----------------------------")

CI_Cat ['subapplication' 'application' 'computer' nan 'displaydevice' 'software'
 'storage' 'database' 'hardware' 'officeelectronics' 'networkcomponents'
 'applicationcomponent' 'Phone']
application             32900
subapplication           7782
computer                 3643
storage                   703
hardware                  442
software                  333
database                  214
displaydevice             212
officeelectronics         152
networkcomponents         107
applicationcomponent        5
Phone                       2
Name: CI_Cat, dtype: int64
----------------------------
CI_Subcat ['Web Based Application' 'Desktop Application' 'Server Based Application'
 'SAP' 'Client Based Application' 'Citrix' 'Standard Application'
 'Windows Server' 'Laptop' 'Linux Server' nan 'Monitor'
 'Automation Software' 'SAN' 'Banking Device' 'Desktop' 'Database'
 'Oracle Server' 'Keyboard' 'Printer' 'Exchange' 'System Software' 'VDI'
 'Encryption' 'Omgeving' 'MigratieDummy' 'Scanner' 

In [34]:
num_data=data.select_dtypes(include=["float64","int64"]).columns
num_data

Index(['Priority', 'number_cnt', 'No_of_Reassignments',
       'No_of_Related_Interactions', 'No_of_Related_Incidents',
       'No_of_Related_Changes'],
      dtype='object')

In [35]:
data.describe().T

count      mean       std       min       25%  \
Priority                    45226.0  4.215805  0.705624  1.000000  4.000000   
number_cnt                  46606.0  0.499658  0.288634  0.000023  0.248213   
No_of_Reassignments         46605.0  1.131831  2.269774  0.000000  0.000000   
No_of_Related_Interactions  46492.0  1.149897  2.556338  1.000000  1.000000   
No_of_Related_Incidents      1222.0  1.669394  3.339687  1.000000  1.000000   
No_of_Related_Changes         560.0  1.058929  0.403596  1.000000  1.000000   

                                 50%       75%         max  
Priority                    4.000000  5.000000    5.000000  
number_cnt                  0.500269  0.749094    0.999997  
No_of_Reassignments         0.000000  2.000000   46.000000  
No_of_Related_Interactions  1.000000  1.000000  370.000000  
No_of_Related_Incidents     1.000000  1.000000   63.000000  
No_of_Related_Changes       1.000000  1.000000    9.000000

In [36]:
data.describe(include="object")

CI_Cat                 CI_Subcat  Status Impact  Urgency  \
count         46495                     46495   46606  46606    46606   
unique           12                        64       2      6       11   
top     application  Server Based Application  Closed      4        4   
freq          32900                     18811   46597  22556    15526   

        Category Alert_Status         Open_Time       Reopen_Time  \
count      46606        46606             46606              2284   
unique         4            1             34636              2244   
top     incident       closed  24-03-2014 08:54  29-10-2013 15:03   
freq       37748        46606                 7                 2   

           Resolved_Time       Close_Time Handle_Time_hrs Closure_Code  
count              44826            46606           46605        46146  
unique             33627            34528           30638           14  
top     10/10/2013 12:53  2/10/2013 15:20               0        Other  
freq                   7               21             236        16470

##Unique value will be 1 for constant column.Here, Alert_Status is the constant feature. so we drop the column.

In [37]:
data.isnull().sum()

CI_Cat                          111
CI_Subcat                       111
Status                            0
Impact                            0
Urgency                           0
Priority                       1380
number_cnt                        0
Category                          0
Alert_Status                      0
No_of_Reassignments               1
Open_Time                         0
Reopen_Time                   44322
Resolved_Time                  1780
Close_Time                        0
Handle_Time_hrs                   1
Closure_Code                    460
No_of_Related_Interactions      114
No_of_Related_Incidents       45384
No_of_Related_Changes         46046
dtype: int64

In [38]:
data.drop(["Alert_Status"],axis=1)

CI_Cat                 CI_Subcat  Status Impact Urgency  \
0      subapplication     Web Based Application  Closed      4       4   
1         application     Web Based Application  Closed      3       3   
2         application       Desktop Application  Closed     NS       3   
3         application     Web Based Application  Closed      4       4   
4         application     Web Based Application  Closed      4       4   
...               ...                       ...     ...    ...     ...   
46601     application  Server Based Application  Closed      4       4   
46602     application  Server Based Application  Closed      4       4   
46603        computer                    Laptop  Closed      5       5   
46604     application     Web Based Application  Closed      4       4   
46605        hardware       DataCenterEquipment  Closed      3       3   

       Priority  number_cnt                 Category  No_of_Reassignments  \
0           4.0    0.601292                 incident                 26.0   
1           3.0    0.415050                 incident                 33.0   
2           NaN    0.517551  request for information                  3.0   
3           4.0    0.642927                 incident                 13.0   
4           4.0    0.345258                 incident                  2.0   
...         ...         ...                      ...                  ...   
46601       4.0    0.231896                 incident                  0.0   
46602       4.0    0.805153                 incident                  0.0   
46603       5.0    0.917466                 incident                  0.0   
46604       4.0    0.701278                 incident                  0.0   
46605       3.0    0.902320                 incident                  0.0   

              Open_Time      Reopen_Time     Resolved_Time        Close_Time  \
0        5/2/2012 13:32              NaN   4/11/2013 13:50   4/11/2013 13:51   
1       12/3/2012 15:44  2/12/2013 12:31   2/12/2013 12:36   2/12/2013 12:36   
2      29-03-2012 12:36              NaN  13-01-2014 15:12  13-01-2014 15:13   
3      17-07-2012 11:49              NaN  14-11-2013 09:31  14-11-2013 09:31   
4       10/8/2012 11:01              NaN   8/11/2013 13:55   8/11/2013 13:55   
...                 ...              ...               ...               ...   
46601  31-03-2014 16:23              NaN  31-03-2014 16:29  31-03-2014 16:29   
46602  31-03-2014 15:03              NaN  31-03-2014 15:29  31-03-2014 15:29   
46603  31-03-2014 15:28              NaN  31-03-2014 15:32  31-03-2014 15:32   
46604  31-03-2014 15:35              NaN  31-03-2014 15:42  31-03-2014 15:42   
46605  31-03-2014 17:24              NaN  31-03-2014 22:47  31-03-2014 22:47   

      Handle_Time_hrs                  Closure_Code  \
0      3,87,16,91,111                         Other   
1      4,35,47,86,389                      Software   
2      4,84,31,19,444  No error - works as designed   
3      4,32,18,33,333                Operator error   
4      3,38,39,03,333                         Other   
...               ...                           ...   
46601           0,095                         Other   
46602     0,428333333                    User error   
46603     0,071666667                      Hardware   
46604     0,116944444                      Software   
46605     0,586388889                      Hardware   

       No_of_Related_Interactions  No_of_Related_Incidents  \
0                             1.0                      2.0   
1                             1.0                      1.0   
2                             1.0                      NaN   
3                             1.0                      NaN   
4                             1.0                      NaN   
...                           ...                      ...   
46601                         1.0                      NaN   
46602                         1.0                      NaN   
46603                

##**Basic Checks Report:**
###**Overview:**
When working with IT Service Management (ITSM) data or any dataset, it's essential to perform basic data checks to ensure the quality and integrity of the data. These basic checks help identify issues early on and ensure that your analysis or project is based on reliable data.

**Data Shape**: We Check the dimensions of the dataset, i.e., the number of rows and columns, using data.shape. Data Contains 46606 Rows and 25 columns.
**Data Types:** We Verified that the data types of each column are as expected usingdata.dtypes.Priority,number_cnt,No_of_Related_Interactions,No_of_Related_Incidents,No_of_Related_Changes,No_of_Reassignments are in float64 and remaining columns are in object.

**Descriptive Statistics:** We Used data.describe().T to get summary statistics (mean, min, max, etc.) for numerical columns.
Use data.describe(include='O').T for a summary of categorical columns.

**Unique Values:** Checked the number of unique values in categorical columns using data['column'].unique(). here, One constant columns are there. so we dropped that column.

**Value Counts:** Examined the distribution of categorical values using df['column'].value_counts().

The goal is to ensure that our data is clean, complete, and suitable for analysis or further processing.

#**Exploratory Data Analysis**